doc/writing-ppxs.mld

```diff
@@ -0,0 +1,510 @@
+
+{0 Writing a Transformation}
+
+This chapter covers the [ppxlib] procedure basics to define and register a
+transformation, be it a global or a context-free transformation.
+
+For the actual manipulation and generation of code, [ppxlib] provides many helpers
+that are listed in {!generatingcode}.
+
+{1 Defining a Transformation}
+
+For [ppxlib], a transformation is a description of a way to modify a given AST
+into another one. A transformation can be:
+
+- A global transformation, which takes the simple form of a function of type
+  [structure -> structure] or [signature -> signature], that can sometimes take
+  extra information as additional arguments. Such a transformation is applied in
+  the {{!driver."global-transfo-phase"}global transformation phase}, unless it
+  has a good reason to have been registered in another phase.
+- A context-free transformation, which only acts on a portion of the AST. In the [ppxlib] framework, those transformations
+  are represented by values of type {{!Ppxlib.Context_free.Rule.t}[Context_free.Rule.t]} and are executed in the {{!driver."context-free-phase"}context-free phase}.
+
+Whenever possible, a context-free transformation should be preferred: global transformations have many {{!global_transformation}drawbacks}, while context-free ones have many {{!driver.advantages}advantages}.
+
+In order to register a transformation to the [ppxlib] driver, one should use the
+{{!Ppxlib.Driver.V2.register_transformation}[Driver.V2.register_transformation]}. This function is used to register all
+rewriter types in every different phase, except derivers, which are abstracted
+away in {{!Ppxlib.Deriving}[Deriving]}.
+
+{1 Context-Free Transformation}
+
+In [ppxlib], the type for context-free transformation is
+{{!Ppxlib.Context_free.Rule.t}[Context_free.Rule.t]}. Rules will be applied during the AST's top-down traverse
+of the context-free pass. A rule contains the information about
+when it should be applied in the traversal, as well as the transformation to
+apply.
+
+Currently, rules can only be defined to apply in four different contexts:
+
+- on extensions points, such as [\[%ext_point payload\]]
+- on some structure or signature items with an attribute, such as
+  [type t = Nil \[@@deriving show\]],
+- on
+  {{:https://v2.ocaml.org/manual/extensionsyntax.html#ss:extension-literals}
+  litterals with modifiers}, such as [41g] or [43.2x],
+- on function application or identifiers, such as [meta_function "99"] and [meta_constant].
+
+In order to define rules on extensions points, we will use the {{!Ppxlib.Extension}[Extension]}
+module. In order to define rules on attributed items, we will use the
+{{!Ppxlib.Deriving}[Deriving]} module. For the two other rules, we will directly use the
+{{!Ppxlib.Context_free.Rule}[Context_free.Rule]} module.
+
+{2 Extenders}
+
+An {{!driver.def_extenders}extender} is characterised by several things:
```

Comment:
This reference seems broken in https://choum.net/panglesd/ppxlib_new_manual/ppxlib/writing-ppxs.html#extenders

Reply:
yes, I did not kept everything up to date as I made changes too often, but you can have a look at the up to date doc by doing
```
$ dune build @doc 
$ firefox _build/default/_doc/_html/index.html
```
(if you are using firefox :wink:)